Assignment.Tests/PingProcessTests.cs

```diff
@@ -1,152 +1,190 @@
+    [DataRow("localhost")]
+    [DataRow("8.8.8.8")]
+    [DataRow("facebook.com")]
+    public void RunAsync_UsingTaskReturn_Success(string address)
+    {
+        //Arrange
+
+        //Act
+
+        Task<PingResult> pingResult = Sut.RunAsync(address);
+        pingResult.Wait();
+        //Assert
+        AssertValidPingOutput(pingResult.Result);
+
+    }
+
+    [TestMethod]
+    [DataRow("www.google.com")]
+    [DataRow("localhost")]
+    [DataRow("8.8.8.8")]
+    [DataRow("facebook.com")]
+    public async Task RunAsync_UsingTpl_Success(string hostName)
+    {
+        // Arrange & Act
+        Task<PingResult> pingResult = Sut.RunAsync(hostName);
+        //Assert
+        AssertValidPingOutput(await pingResult);
+    }
+
+    [TestMethod]
+    [ExpectedException(typeof(AggregateException))]
+    public void RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping()
+    {
+        // Arrange
+        CancellationToken cancellationToken = new(true);
+
+        // Act
+        Task<PingResult> pingResult = Sut.RunAsync("localhost", cancellationToken);
+        pingResult.Wait();
+    }
+
+    [TestMethod]
+    [ExpectedException(typeof(TaskCanceledException))]
+    public void RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException()
+    {
+        // Use exception.Flatten()
+        // Arrange
+        CancellationToken cancellationToken = new(true);
+
+        // Act
+        Task<PingResult> pingResult = Sut.RunAsync("localhost", cancellationToken);
+
+        // Assert
+        Assert.ThrowsException<AggregateException>(pingResult.Wait).Flatten().Handle((exception) => throw exception);
+    }
+
+    [TestMethod]
+    public async Task RunAsync_MultipleHostAddresses_True()
+    {
+        // Pseudo Code - don't trust it!!!
+
+        // Arrange
+        string[] hostNames = ["localhost", "localhost", "localhost", "localhost"];
+        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length * hostNames.Length;
+        PingResult result = await Sut.RunAsync(hostNames);
+
+        // Act
+        int? lineCount = result.StdOutput?.Split(Environment.NewLine).Length;
+
+        // Assert
+        Assert.AreEqual(0, result.ExitCode);
+        Assert.AreEqual(expectedLineCount, lineCount);
+    }
+
+    [TestMethod]
+
+    public async Task RunLongRunningAsync_UsingTpl_Success()
+    {
+        PingResult result = await Sut.RunLongRunningAsync("localhost");
+        AssertValidPingOutput(result);
+
+    }
+
+    //[TestMethod]
+    //public void StringBuilderAppendLine_InParallel_IsNotThreadSafe()
+    //{
+    //    IEnumerable<int> numbers = Enumerable.Range(0, short.MaxValue);
+    //    System.Text.StringBuilder stringBuilder = new();
+    //    numbers.AsParallel().ForAll(item => stringBuilder.AppendLine(""));
+    //    int lineCount = stringBuilder.ToString().Split(Environment.NewLine).Length;
+    //    Assert.AreNotEqual(lineCount, numbers.Count() + 1);
+    //}
```

Comment:
Use Try - catch and it will work:

```suggestion
  try
        {
            IEnumerable<int> numbers = Enumerable.Range(0, short.MaxValue);
            System.Text.StringBuilder stringBuilder = new();
            numbers.AsParallel().ForAll(item => stringBuilder.AppendLine(""));
            int lineCount = stringBuilder.ToString().Split(Environment.NewLine).Length;
            Assert.AreNotEqual(lineCount, numbers.Count() + 1);
        }
        catch(AggregateException) { 
            //Ignore Destination too short
        }```

Reply:
That's true, for now we will leave it commented out, but catching it is a good idea as well!